docs/rfcs/1005-realm-key-rotation.md

```diff
@@ -0,0 +1,924 @@
+
+### 6.5 - Automatic key rotation
+
+For automatic key rotation, a new "key rotation manager" is to be introduced. Its function
+is very simple:
+
+1. Wait for new `UserRevokedCertificate`, `RealmRoleCertificate` and
+   `RealmKeyRotationCertificate` certificates
+2. For `UserRevokedCertificate`, call `realm_unshare` for each realm we are owner (and
+   the revoked user is part of)
+3. For `RealmRoleCertificate`, if the role is None and we are owner of the realm:
+    1. Wait a random time between 30 and 60 seconds (and reset the wait if we were
+       already waiting for this realm)
+    2. Call `realm_rotate_key`
+4. For `RealmKeyRotationCertificate` cancel the wait from step 3.1 if any
+
+The wait at step 3.1 is here for two reasons:
+
+- Give priority to the author of the certificates to do the key rotation. This is
+  typically needed if he is doing un-sharing batch (and hence will do the key rotation
+  only once all un-sharing have been done)
+- Avoid all owners to send key rotation command at the same time to the server.
+
+### 6.6 - Existing local data: `LocalUserManifest`
+
+`LocalUserManifest` has to be modified to reflect changes in `UserManifest`:
+
+- remove `last_processed_message` field
+- change `workspace` field to use a updated `WorkspaceEntry` type
+
+```json
+{
+  "label": "WorkspaceEntry",
+  "fields": [
+    {
+      "name": "id",
+      "type": "VlobID"
+    },
+    {
+      // `name` used to be a non-optional `EntryName` synchronized field.
+      // Now it has become option (msgpack encoding makes this change backward compatible)
+      // and is a local only field (i.e. never synchronized) that is used to store the
+      // name of the workspace on creation.
+      // This only needed in the brief moment before the `RealmNameCertificate` is uploaded
+      // (which can take some time if e.g. the workspace is created while the client is offline)
+      "name": "name",
+      "type": "RequiredOption<EntryName>"
+    }
+    // Legacy ignored fields
+    // - `key`
+    // - `encryption_revision`
+    // - `encrypted_on`
+    // - `role_cache_timestamp`
+    // - `role`
+  ]
+}
+```
+
+### 6.7 - Add HTTP `User-Agent` header
+
+Currently the client send no `User-Agent` header to the server.
+
+Instead we should set it `Parsec-client/<version>`. (e.g. `Parsec-Client/3.0.1`).
+
+This allows us to use reverse proxy in front of the SAAS server to
+[deal with Parsec v2 vs Parsec v3 client](#71---server-side-organization-migration)
+
+## 7 - Backward compatibility & migration
+
+As stated, we don't support APIv3 and APIv4 in the same Parsec server.
+
+An APIv3 Parsec server can be migrate to APIv4.
+
+After that it will deny any APIv3 request, hence forcing all it users to upgrade
+their Parsec client at once.
+
+### 7.1 - Server-side organization migration
+
+This is typically for the SAAS version Parsec: some clients may want to be able
+to control when the switch from v2 to v3 should come for their organization.
```

Comment:
```suggestion
This is typically for the SAAS version Parsec: 

Some clients may want to be able to control when the switch from v2 to v3 should come for their organization.
```

Reply:
I don't think this is needed: the sentence is already small enough.